File: tnzhou/213.cpp
class Solution {
public:
    int rob(vector<int>& nums) {
        int last_house_profit = 0;
        int second_last_profit = 0;
        int curr_profit = 0;
        int max_profit = 0;
        if (nums.size() == 0)
            return 0;
        if (nums.size() == 1) 
            return nums[0]; // if only one house, directly rob it
        for (int i = 0; i < nums.size() - 1; i++) {
            curr_profit = max(
                last_house_profit,
                second_last_profit + nums[i]
            );
            max_profit = curr_profit > max_profit ? curr_profit : max_profit;
            second_last_profit = last_house_profit;
            last_house_profit = curr_profit;
        }
        last_house_profit = 0;
        second_last_profit = 0;
        curr_profit = 0;
        for (int i = 1; i < nums.size(); i++) {
            curr_profit = max(
                last_house_profit,
                second_last_profit + nums[i]
            );
            max_profit = curr_profit > max_profit ? curr_profit : max_profit;
            second_last_profit = last_house_profit;
            last_house_profit = curr_profit;
        }
        return max_profit;
    }
};

File: chunlin/213.java
class Solution {
    public int rob(int[] nums) {
        if (nums == null || nums.length == 0) {
            return 0;
        }
        if (nums.length == 1) return nums[0];
        int n = nums.length;
        return Math.max(helper(0, n - 2, nums), helper(1, n - 1, nums));
    }
    public int helper(int lo, int hi, int[] nums) {
        int[] opt = new int[hi - lo + 2];
        opt[0] = 0;
        opt[1] = nums[lo];
        for (int i = 2; i <= hi - lo + 1; i++) {
            opt[i] = Math.max(opt[i - 1], opt[i - 2] + nums[lo + i - 1]);
        }
        return opt[hi - lo + 1];
    }
}
